### PR TITLE
Jsr311 path param processing

### PR DESCRIPTION
**Pull request by iainduncani**
Fix for issue #363

### CODE REVIEW

File: curly.go [hunk 1]
```diff
@@ -18,7 +19,7 @@ type CurlyRouter struct{}
 // for the WebService and its Route for the given Request.
 func (c CurlyRouter) SelectRoute(
 	webServices []*WebService,
-	httpRequest *http.Request) (selectedService *WebService, selected *Route, err error) {
+	httpRequest *http.Request) (selectedService *WebService, selected *Route, params map[string]string, err error) {
```
Comment (emicklei) [inline curly.go]: this is a breaking change as the exposed function signature is changed.  was this really needed?  I will continue reviewing.

File: curly_test.go [hunk 2]
```diff
@@ -158,6 +158,63 @@ func TestExtractParameters_Wildcard3(t *testing.T) {
 	}
 }
 
+func TestExtractParameters_Wildcard4(t *testing.T) {
```
Comment (emicklei) [inline curly_test.go]: why did you move the function definition to another file?

File: curly.go [hunk 3]
```diff
@@ -162,3 +168,46 @@ func (c CurlyRouter) computeWebserviceScore(requestTokens []string, tokens []str
 	}
 	return true, score
 }
+
+// Extract the parameters from the request url path
+func extractParameters(r *Route, urlParts []string) map[string]string {
```
Comment (emicklei) [inline curly.go]: why did you move this definition to another file?

File: curly.go (hunk 1 above)
Comment (iainduncani) [reply curly.go]: Ah, I hadn't appreciated that other people might supply their own router so I thought as long as I updated the two in here this would be OK.  I'd done this as I wanted the router to be responsible for doing the calculation of the parameters as it's them that changes how they do routing and handling of the params.  Instead I think it's probably better to do this through an extra interface that the routers can optionally implement, it's not quite as clean but better than introducing a breaking change...

File: curly.go (hunk 3 above)
Comment (iainduncani) [reply curly.go]: To put it into the router as different routers treat params differently in their routing handling, as with the comment above I'll use this default behaviour if the router doesn't calculate the params itself

**Comment by iainduncani** [comment]
Closed to make extra changes that don't introduce breaking change
